### PR TITLE
Issue#243 high cpu issue of CsdlSemanticsModel

### PR DESCRIPTION
**Pull request by karataliu**

### CODE REVIEW

**Comment by congysu** [comment]
#243

File: ODataLib/EdmLib/Desktop/.Net4.0/Microsoft/Data/Edm/Internal/EdmUtil.cs [hunk 1]
```diff
@@ -294,6 +294,39 @@ public static bool EqualsOrdinalIgnoreCase(this string string1, string string2)
...
+
+            TValue val;
+            if (!dictionary.TryGetValue(key, out val))
+            {
+                lock (dictionary)
+                {
+                    if (!dictionary.TryGetValue(key, out val))
+                    {
+                        val = computeValue(key);
+                        dictionary.Add(key, val);
```
Comment (congysu) [inline ODataLib/EdmLib/Desktop/.Net4.0/Microsoft/Data/Edm/Internal/EdmUtil.cs]: Make sure this is the only place to add an item.

File: ODataLib/EdmLib/Desktop/.Net4.0/Microsoft/Data/Edm/Csdl/Internal/Semantics/CsdlSemanticsModel.cs [hunk 2]
```diff
@@ -451,22 +451,17 @@ internal IEnumerable<IEdmVocabularyAnnotation> WrapInlineVocabularyAnnotations(C
...
-
-            CsdlValueAnnotation valueAnnotation = annotation as CsdlValueAnnotation;
-            result =
-                valueAnnotation != null
-                ? (CsdlSemanticsVocabularyAnnotation)new CsdlSemanticsValueAnnotation(schema, targetContext, annotationsContext, valueAnnotation, qualifier)
-                : (CsdlSemanticsVocabularyAnnotation)new CsdlSemanticsTypeAnnotation(schema, targetContext, annotationsContext, (CsdlTypeAnnotation)annotation, qualifier);
-
-            this.wrappedAnnotations[annotation] = result;
-            return result;
+            return EdmUtil.DictionaryGetOrUpdate(
```
Comment (congysu) [inline ODataLib/EdmLib/Desktop/.Net4.0/Microsoft/Data/Edm/Csdl/Internal/Semantics/CsdlSemanticsModel.cs]: Shall we also fix this for OData v4.

**Comment by congysu** [comment]
See if this can fit in 5.6.5, or 5.6.6.

**Comment by karataliu** [comment]
Add https://github.com/OData/odata.net/issues/289 to track porting this to v4.

**Pull request closed without merging**